PROTOCOL.md

```diff
@@ -0,0 +1,51 @@
+
+## Client Layer
+
+Clients connect to an arbitrary node though a GRPC based [message API](https://github.com/xmtp/proto/blob/main/proto/message_api/v1/message_api.proto) and interact with the network as a whole through this node. The message API is the interface separating the client layer from the node layer. Conversations are represented at the node layer as topics.
+
+The client layer uses protobuf for encoding of all its structures. The definitions are maintained in the [xtmp/proto repository](https://github.com/xmtp/proto/blob/main/proto).
+
+### Keys
+
+A client is associated with a wallet address through a set of public/private keys (EC secp256k1) generated when the wallet is registered with the XMTP network. First there is the identity key that serves as a proxy for the wallet key and is signed by the wallet to establish its authenticity. Then there is one or more pre-keys that are used for message encryption. The pre-key authenticity is established with an identity key signature. The two level key structure is there to facilitate more frequent key rotation without requiring frequent wallet signing. The key structure and naming is modelled after [Signal's X3DH protocol](https://signal.org/docs/specifications/x3dh/#the-x3dh-protocol).
+
+The keys are maintained and stored together in a key bundle, the private keys in a [PrivateKeyBundle](https://github.com/xmtp/proto/blob/main/proto/message_contents/private_key.proto) and public keys in a [PublicKeyBundle](https://github.com/xmtp/proto/blob/main/proto/message_contents/public_key.proto).
+
+### Contacts
+
+To start a conversation the initiator needs the public keys of the participants. To facilitate this step registered clients publish their public key bundle in easily discoverable contact topics.
+
+Each user is automatically assigned a contact topic named `contact-<user's wallet address>`. Public key bundles are published as [ContactBundles](https://github.com/xmtp/proto/blob/main/proto/message_contents/contact.proto) which allow for versioning of the bundle contents and format. A public key bundle MUST be republished whenever the pre-key or identity-key changes. The latest published bundle represents the current bundle that should be used for all new messages and conversations.
+
+(TODO) When publishing a contact into a contact topic, the nodes MUST verify that the bundle being published is a valid bundle signed by the wallet that owns the contact topic (i.e. the address in the topic name matches the address of the wallet that signed the bundle). The nodes must also verify that the created_ns time of the pre-key in the bundle is newer than the created_ns time of the pre-key in the latest published bundle.
```

Comment:
Hm, the TODO marker is a bit misleading, but I'm looking at the spec as a description of what we think the protocol should be, not necessarily a reflection of the current state of our SDK or node software. Of course we should strive to make our SDK and node compliant with the spec and I assume we will, but to me there is a distinction if that makes sense.

Reply:
I think I'm going to drop this bit after all. It can be brought back as a future revision if we end up agreeing that this is a good idea.

Reply:
I replaced the paragraph with

> Only account owners SHOULD be able to publish into their contact topics. This could be enforced at the API or at the node level. Specifics are out of scope in this document.

We can revise later as we see fit.